server/sonar-server/src/main/java/org/sonar/server/computation/step/UnitTestMeasuresStep.java

```diff
@@ -0,0 +1,150 @@
+ */
+public class UnitTestMeasuresStep implements ComputationStep {
+
+  private final TreeRootHolder treeRootHolder;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  private final ImmutableList<Formula> FORMULAS = ImmutableList.<Formula>of(
+    new SumFormula(TEST_EXECUTION_TIME_KEY),
+    new SumFormula(SKIPPED_TESTS_KEY),
+    new UnitTestsFormula()
+    );
+
+  public UnitTestMeasuresStep(TreeRootHolder treeRootHolder, MetricRepository metricRepository, MeasureRepository measureRepository) {
+    this.treeRootHolder = treeRootHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+  }
+
+  @Override
+  public void execute() {
+    FormulaExecutorComponentVisitor.newBuilder(metricRepository, measureRepository)
+      .buildFor(FORMULAS)
+      .visit(treeRootHolder.getRoot());
+  }
+
+  private static class UnitTestsFormula implements Formula<UnitTestsCounter> {
+
+    @Override
+    public UnitTestsCounter createNewCounter() {
+      return new UnitTestsCounter();
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(UnitTestsCounter counter, CreateMeasureContext context) {
+      return createMeasure(context, counter.testsCounter.getValue(), TESTS_KEY)
+        .or(createMeasure(context, counter.testsErrorsCounter.getValue(), TEST_ERRORS_KEY))
+        .or(createMeasure(context, counter.testsFailuresCounter.getValue(), TEST_FAILURES_KEY))
+        .or(createDensityMeasure(counter, context));
+    }
+
+    private static Optional<Measure> createMeasure(CreateMeasureContext context, Optional<Integer> metricValue, String metricKey) {
+      if (context.getMetric().getKey().equals(metricKey) && metricValue.isPresent() && context.getComponent().getType().isHigherThan(Component.Type.FILE)) {
+        return Optional.of(Measure.newMeasureBuilder().create(metricValue.get()));
+      }
+      return Optional.absent();
+    }
+
+    private static Optional<Measure> createDensityMeasure(UnitTestsCounter counter, CreateMeasureContext context) {
+      if (context.getMetric().getKey().equals(TEST_SUCCESS_DENSITY_KEY)
+        && isPositive(counter.testsCounter.getValue(), true)
+        && isPositive(counter.testsErrorsCounter.getValue(), false)
+        && isPositive(counter.testsFailuresCounter.getValue(), false)) {
+        int tests = counter.testsCounter.getValue().get();
+        int errors = counter.testsErrorsCounter.getValue().get();
+        int failures = counter.testsFailuresCounter.getValue().get();
+        double density = (errors + failures) * 100d / tests;
+        return Optional.of(Measure.newMeasureBuilder().create(100d - density));
+      }
+      return Optional.absent();
+    }
+
+    private static boolean isPositive(Optional<Integer> value, boolean isStrictComparison) {
+      return value.isPresent() && (isStrictComparison ? value.get() > 0 : value.get() >= 0);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)